impeller/renderer/backend/vulkan/gpu_tracer_vk.cc

```diff
@@ -5,127 +5,155 @@
+bool GPUTracerVK::IsValid() const {
+  return valid_;
+}
+
+void GPUTracerVK::MarkFrameStart() {
+  in_frame_ = true;
+}
+
+void GPUTracerVK::MarkFrameEnd() {
   if (!valid_) {
     return;
   }
-  auto strong_context = context_.lock();
-  if (!strong_context) {
-    return;
-  }
-  auto buffer = strong_context->CreateCommandBuffer();
-  auto vk_trace_cmd_buffer =
-      CommandBufferVK::Cast(*buffer).GetEncoder()->GetCommandBuffer();
-  // The two commands below are executed in order, such that writeTimeStamp is
-  // guaranteed to occur after resetQueryPool has finished. The validation
-  // layer seem particularly strict, and efforts to reset the entire pool
-  // were met with validation errors (though seemingly correct measurements).
-  // To work around this, the tracer only resets the query that will be
-  // used next.
-  vk_trace_cmd_buffer.resetQueryPool(query_pool_.get(), current_index_, 1);
-  vk_trace_cmd_buffer.writeTimestamp(vk::PipelineStageFlagBits::eBottomOfPipe,
-                                     query_pool_.get(), current_index_);
-
-  if (!buffer->SubmitCommands()) {
-    VALIDATION_LOG << "GPUTracerVK: Failed to record start time.";
-  }
+  Lock lock(trace_state_mutex_);
+  current_state_ = (current_state_ + 1) % 16;
+
+  auto& state = trace_states_[current_state_];
+  FML_DCHECK(state.pending_buffers == 0u);
 
-  // The logic in RecordEndFrameTime requires us to have recorded a pair of
-  // tracing events. If this method failed for any reason we need to be sure we
-  // don't attempt to record and read back a second value, or we will get values
-  // that span multiple frames.
-  started_frame_ = true;
+  state.pending_buffers = 0;
+  state.current_index = 0;
+  state.contains_failure = false;
+  in_frame_ = false;
 }
 
-void GPUTracerVK::RecordEndFrameTime() {
-  if (!valid_ || !started_frame_) {
+void GPUTracerVK::RecordCmdBufferStart(const vk::CommandBuffer& buffer) {
+  if (!valid_ || !in_frame_) {
     return;
   }
-  auto strong_context = context_.lock();
-  if (!strong_context) {
+  Lock lock(trace_state_mutex_);
+  auto& state = trace_states_[current_state_];
+
+  // Initialize the query pool for the first query on each frame.
+  if (state.pending_buffers == 0) {
+    vk::QueryPoolCreateInfo info;
+    info.queryCount = kPoolSize;
+    info.queryType = vk::QueryType::eTimestamp;
+
+    auto [status, pool] =
+        device_holder_->GetDevice().createQueryPoolUnique(info);
+    if (status != vk::Result::eSuccess) {
+      VALIDATION_LOG << "Failed to create query pool.";
+      return;
+    }
+    trace_states_[current_state_].query_pool = std::move(pool);
+    buffer.resetQueryPool(trace_states_[current_state_].query_pool.get(), 0,
+                          kPoolSize);
+  }
+
+  // We size the query pool to kPoolSize, but Flutter applications can create an
+  // unbounded amount of work per frame. If we encounter this, stop recording
+  // cmds.
+  if (state.current_index >= kPoolSize) {
```

Comment:
Do you expect this to be common?

I wonder if it makes sense to at least log that recording did not happen.

Reply:
No, I think this would be extremely rare but adapting the tracer to work with arbitrarily resizing query pools is much harder so punting for now.